File: src/main/java/de/itsTyrion/flusenbot/command/CommandHandler.java
package de.itsTyrion.flusenbot.command;

import com.pengrad.telegrambot.TelegramBot;
import com.pengrad.telegrambot.model.Chat;
import com.pengrad.telegrambot.model.Message;
import com.pengrad.telegrambot.model.User;
import com.pengrad.telegrambot.request.DeleteMessage;
import com.pengrad.telegrambot.request.GetChatMember;
import com.pengrad.telegrambot.request.SendMessage;
import de.itsTyrion.flusenbot.cache.AdminCache;
import de.itsTyrion.flusenbot.command.commands.Cooldown;
import de.itsTyrion.flusenbot.command.commands.Ping;
import de.itsTyrion.flusenbot.command.commands.ReloadAdmins;
import de.itsTyrion.flusenbot.command.commands.Soon;
import de.itsTyrion.flusenbot.util.Utils;
import lombok.AllArgsConstructor;
import lombok.NonNull;

import java.util.Arrays;
import java.util.Map;

@AllArgsConstructor
public final class CommandHandler {
    private final @NonNull TelegramBot api;
    private final Map<String, Command> commands = Map.of(
            "ping", new Ping(),
            "reloadadmins", new ReloadAdmins(),
            "cooldown", new Cooldown(),
            "kick", new Soon(),
            "mute", new Soon(),
            "ban", new Soon()
    );

    public static final int ID_TYRION = 218446038;

    public void handleCommand(@NonNull Message msg, @NonNull Chat chat, @NonNull User user) {
        var text = msg.text();
        if (text.contains("@") && !text.contains("flusenbot"))
            return;

        var split = text.split("\\s+"); // Pattern mathing one or more spaces
        var command = commands.get(split[0].substring(1).replace("@flusenbot", "").toLowerCase());

        if (command == null)
            return;

        boolean hasPermissions = true;
        if (command.getPermissions().length != 0) {
            var member = api.execute(new GetChatMember(chat.id(), user.id())).chatMember();
            hasPermissions = Arrays.stream(command.getPermissions()).allMatch(permission -> permission.check(member))
                    || command.getName().equals("cooldown") && chat.id() == -1001238995053L && user.id() == ID_TYRION;
        }

        if ((!command.isAdminOnly() || AdminCache.isAdmin(user, chat)) && hasPermissions) {
            command.execute(Arrays.copyOfRange(split, 1, split.length), chat, user);
        } else {
            int id = api.execute(new SendMessage(chat.id(), "⚠️No permission")).message().messageId();
            Utils.runDelayed(() -> api.execute(new DeleteMessage(chat.id(), id)), 5);
        }
    }
}

File: src/main/java/de/itsTyrion/flusenbot/command/CommandHandler.kt
package de.itsTyrion.flusenbot.command

import com.pengrad.telegrambot.TelegramBot
import com.pengrad.telegrambot.model.Chat
import com.pengrad.telegrambot.model.Message
import com.pengrad.telegrambot.model.User
import com.pengrad.telegrambot.request.DeleteMessage
import com.pengrad.telegrambot.request.GetChatMember
import com.pengrad.telegrambot.request.SendMessage
import de.itsTyrion.flusenbot.cache.AdminCache.isAdmin
import de.itsTyrion.flusenbot.command.commands.*
import de.itsTyrion.flusenbot.util.Utils.runDelayed

class CommandHandler(private val api: TelegramBot) {
    private val commands = mapOf(
            "ping" to Ping(),
            "reloadadmins" to ReloadAdmins(),
            "cooldown" to Cooldown(),
            "kick" to Kick(),
            "mute" to Soon(),
            "ban" to Soon()
    )

    fun handleCommand(msg: Message, chat: Chat, user: User) {
        val text = msg.text()
        if (text.contains("@") && !text.contains("flusenbot"))
            return

        val split = text.split("\\s+").toTypedArray() // Pattern mathing one or more spaces
        val command = commands[split[0].substring(1).replace("@flusenbot", "").lowercase()] ?: return
        var hasPermissions = true

        if (command.permissions.isNotEmpty()) {
            val member = api.execute(GetChatMember(chat.id(), user.id())).chatMember()

            println("check ${member.user().username()}")
            println(member.status())
            Command.Permission.values().forEach {
                try {
                    println(it.check(member))
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
            hasPermissions = (command.permissions.all { it.check(member) }
                    || command.name == "cooldown" && chat.id() == -1001238995053L && user.id() == ID_TYRION)
        }
        if ((!command.isAdminOnly || isAdmin(user, chat)) && hasPermissions) {
            command.execute(split.copyOfRange(1, split.size), msg)
        } else {
            val id = api.execute(SendMessage(chat.id(), "⚠️No permission")).message().messageId()
            runDelayed({ api.execute(DeleteMessage(chat.id(), id)) }, 5)
        }
    }

    companion object {
        const val ID_TYRION = 218446038L
    }
}